### PR TITLE
Don't throw exception when trying to resolve a possible link path

### LINKED ISSUE #16293: OperationStopped: Incorrect function when running an exe from a mounted ISO
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

My repro steps:
- Download the Enterprise WDK: https://docs.microsoft.com/en-us/windows-hardware/drivers/other-wdk-downloads
- Mount the ISO (Mount-DiskImage -ImagePath S:\iso\dev\EWDK_fe_release_20348_210507-1500.iso).
- From a "cmd.exe" prompt, run LaunchBuildEnv.cmd amd64
- Start Powershell 7.2 RC1: "c:\Program Files\PowerShell\7-preview\pwsh.exe"
- Run MSBuild.exe

But I do _not_ think this is linked to the EWDK but the fact I'm running a exe from a mounted iso.
(was running just fine with 7.1.5 or the previous 7.2 preview)

### Expected behavior

```console
Successfully run msbuild.exe
```

### Actual behavior

```console
OperationStopped: Incorrect function. : 'E:\Program Files\Microsoft Visual Studio\2019\BuildTools\\MSBuild\Current\Bin\MSBuild.exe'
```

### Error details

```console
Exception             :
    Type       : System.IO.IOException
    TargetSite :
        Name          : GetImmediateLinkTarget
        DeclaringType : System.IO.FileSystem, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral,
PublicKeyToken=7cec85d7bea7798e
        MemberType    : Method
        Module        : System.Private.CoreLib.dll
    Message    : Incorrect function. : 'E:\Program Files\Microsoft Visual
Studio\2019\BuildTools\\MSBuild\Current\Bin\MSBuild.exe'
    Data       : System.Collections.ListDictionaryInternal
    Source     : System.Private.CoreLib
    HResult    : -2147024895
    StackTrace :
   at System.IO.FileSystem.GetImmediateLinkTarget(String linkPath, Boolean isDirectory, Boolean throwOnError, Boolean
returnFullPath)
   at System.IO.FileSystem.ResolveLinkTarget(String linkPath, Boolean returnFinalTarget, Boolean isDirectory)
   at System.IO.File.ResolveLinkTarget(String linkPath, Boolean returnFinalTarget)
   at System.Management.Automation.NativeCommandProcessor.IsWindowsApplication(String fileName)
   at System.Management.Automation.NativeCommandProcessor.InitNativeProcess()
   at System.Management.Automation.NativeCommandProcessor.Prepare(IDictionary psDefaultParameterValues)
   at System.Management.Automation.CommandProcessorBase.DoPrepare(IDictionary psDefaultParameterValues)
   at System.Management.Automation.Internal.PipelineProcessor.Start(Boolean incomingStream)
   at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input)
   at System.Management.Automation.PipelineOps.InvokePipeline(Object input, Boolean ignoreInput,
CommandParameterInternal[][] pipeElements, CommandBaseAst[] pipeElementAsts, CommandRedirection[][] commandRedirections,
FunctionContext funcContext)
   at System.Management.Automation.Interpreter.ActionCallInstruction`6.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
CategoryInfo          : OperationStopped: (:) [], IOException
FullyQualifiedErrorId : System.IO.IOException
InvocationInfo        :
    ScriptLineNumber : 1
    OffsetInLine     : 1
    HistoryId        : -1
    Line             : msbuild.exe
    PositionMessage  : At line:1 char:1
                       + msbuild.exe
                       + ~~~~~~~~~~~
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.2.0-rc.1
PSEdition                      Core
GitCommitId                    7.2.0-rc.1
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

image

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Fix #16293
Before #16165, we were using `InternalSymbolicLinkLinkCodeMethods.WinInternalGetTarget` to try resolving a possible link path, and that API never throws -- catch exception and returns null. However, `File.ResolveLinkTarget` throws when there is any error while resolving the link path, e.g. the file system doesn't support reparse point. This caused a regression in 7.2.0-rc1.

The fix is to wrap `File.ResolveLinkTarget` in a try/catch block, and use the original `fileName` when `ResolveLinkTarget` fails.

@iSazonov and @jborean93, thank you both for bring the regression to our attention, and also worked out the test.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively

### CODE REVIEW

**Review by jborean93 — APPROVED** [review]

**Review by iSazonov — APPROVED** [review]

File: test/powershell/Language/Scripting/NativeExecution/NativeCommandProcessor.Tests.ps1 [hunk 1]
```diff
@@ -252,3 +252,48 @@ Categories=Application;
...
+
+    AfterAll {
+        if ($IsWindows) {
+            Dismount-DiskImage -ImagePath $vhdx
+            Remove-Item $vhdx, $create_vhdx -Force
+        }
+    }
+
+    It "Should run 'whoami.exe' from FAT file system without error" -Skip:(!$IsWindows) {
+        $expected = C:\Windows\System32\whoami.exe
```
Comment (adityapatwardhan) [inline test/powershell/Language/Scripting/NativeExecution/NativeCommandProcessor.Tests.ps1:295]:
```suggestion
        $expected = "$env:WinDir\System32\whoami.exe"
```

File: test/powershell/Language/Scripting/NativeExecution/NativeCommandProcessor.Tests.ps1 [hunk 2]
```diff
@@ -252,3 +252,48 @@ Categories=Application;
...
+            create partition primary
+            format fs=fat
+            assign letter="T"
+            detach vdisk
+"@
+
+        diskpart.exe /s $create_vhdx
+        Mount-DiskImage -ImagePath $vhdx > $null
+
+        Copy-Item C:\Windows\System32\whoami.exe T:\whoami.exe
```
Comment (adityapatwardhan) [inline test/powershell/Language/Scripting/NativeExecution/NativeCommandProcessor.Tests.ps1:284]:
```suggestion
        Copy-Item "$env:WinDir\System32\whoami.exe T:\whoami.exe"
```

**Comment by adityapatwardhan** [comment]
@daxian-dbw New test is failing.. please have a look

**Review by adityapatwardhan — APPROVED** [review]

**Review by t10-13rocket — APPROVED** [review]
LGTM

**Pull request merged**